### PR TITLE
CLOUDSTACK-9570: Bug in listSnapshots for snapshots with deleted data stores

### PR DESCRIPTION
**Pull request by nvazquez**
## Actual behaviour

If there is snapshot on a data store that is removed, `listSnapshots` still tries to enumerate it and gives error (in this example data store 2 has been removed):
`client/api?command=listSnapshots&isrecursive=true&listall=true`

``` xml
<listsnapshotsresponse cloud-stack-version="4.9.1.0-SNAPSHOT">
   <errorcode>530</errorcode>
   <cserrorcode>4250</cserrorcode>
   <errortext>Unable to locate datastore with id 2</errortext>
</listsnapshotsresponse>
```
## Reproduce error

This steps can be followed to reproduce issue:
- Take a snapshot of a volume (this creates a references for primary storage and secondary storage in snapshot_store_ref table
- Simulate retiring primary data storage where snapshot is cached (in this example X is a fake data store and Y is snapshot id):

```
UPDATE `cloud`.`snapshot_store_ref` SET `store_id`='X', `state`="Destroyed" WHERE `id`='Y';
```
- List snapshots: `client/api?command=listSnapshots&isrecursive=true&listall=true`

### CODE REVIEW

**Comment by rhtyd** [comment]
LGTM. I'll fire some regression tests.
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by abhinandanprateek** [comment]
LGTM on code review.

**Comment by serg38** [comment]
@rhtyd @jburwell The tests didn't kick in last time. Can we fire matrix again?

**Comment by ustcweizhou** [comment]
if the data store is deleted, the snapshot should be removed as well, right?

**Comment by rhtyd** [comment]
@blueorangutan package

File: server/src/com/cloud/api/ApiResponseHelper.java [hunk 1]
```diff
@@ -526,16 +529,18 @@ public static DataStoreRole getDataStoreRole(Snapshot snapshot, SnapshotDataStor
         }
 
         long storagePoolId = snapshotStore.getDataStoreId();
-        DataStore dataStore = dataStoreMgr.getDataStore(storagePoolId, DataStoreRole.Primary);
+        if (! snapshotStore.getState().equals(ObjectInDataStoreStateMachine.State.Destroyed)){
```
Comment (rhtyd) [inline server/src/com/cloud/api/ApiResponseHelper.java]: @nvazquez this can cause NPE, instead use `ObjectInDataStoreStateMachine.State.Destroyed.equals(...)` or an explicit != null.

**Comment by serg38** [comment]
@ustcweizhou Not necessary. The issue affects removal of primary storage DS where there are snapshot copies. After that the snapshot could remain on secondary storage.

File: server/src/com/cloud/api/ApiResponseHelper.java (hunk 1 above)
Comment (nvazquez) [inline server/src/com/cloud/api/ApiResponseHelper.java]: Done, thanks @rhtyd

**Comment by jburwell** [comment]
@nvazquez could you please add a Marvin test to verify this fix?

**Comment by nvazquez** [comment]
@jburwell sure, will work on it

File: server/src/com/cloud/api/ApiResponseHelper.java [hunk 2]
```diff
@@ -526,16 +529,18 @@ public static DataStoreRole getDataStoreRole(Snapshot snapshot, SnapshotDataStor
...
 
-        Map<String, String> mapCapabilities = dataStore.getDriver().getCapabilities();
+            Map<String, String> mapCapabilities = dataStore.getDriver().getCapabilities();
 
-        if (mapCapabilities != null) {
-            String value = mapCapabilities.get(DataStoreCapabilities.STORAGE_SYSTEM_SNAPSHOT.toString());
-            Boolean supportsStorageSystemSnapshots = new Boolean(value);
+            if (mapCapabilities != null) {
+                String value = mapCapabilities.get(DataStoreCapabilities.STORAGE_SYSTEM_SNAPSHOT.toString());
+                Boolean supportsStorageSystemSnapshots = new Boolean(value);
```
Comment (jburwell) [inline server/src/com/cloud/api/ApiResponseHelper.java]: `new Boolean` skips the constant pool -- putting unnecessary pressure on the heap and creating a potential memory leak.  Please use `Boolean.valueOf` to part the value to avoid this issue.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@jburwell I added new Marvin test in `test_snapshots.py`

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
LGTM.

**Comment by jburwell** [comment]
@nvazquez could you please check into the Travis failures?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@jburwell the Travis failure is ignorable caused by a oobm corner case (exception is ignorable). I'll work on a fix soon.
@blueorangutan test

**Comment by nvazquez** [comment]
@jburwell @rhtyd sorry I've been off yesterday, do I push force to kick tests again?

File: server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java [hunk 3]
```diff
@@ -517,10 +517,10 @@ public String getSecondaryStorageURL(SnapshotVO snapshot) {
...
 
         Ternary<Long, Boolean, ListProjectResourcesCriteria> domainIdRecursiveListProject = new Ternary<Long, Boolean, ListProjectResourcesCriteria>(cmd.getDomainId(), cmd.isRecursive(), null);
-       _accountMgr.buildACLSearchParameters(caller, id, cmd.getAccountName(), cmd.getProjectId(), permittedAccounts, domainIdRecursiveListProject, cmd.listAll(), false);
-       Long domainId = domainIdRecursiveListProject.first();
-       Boolean isRecursive = domainIdRecursiveListProject.second();
-       ListProjectResourcesCriteria listProjectResourcesCriteria = domainIdRecursiveListProject.third();
+        _accountMgr.buildACLSearchParameters(caller, id, cmd.getAccountName(), cmd.getProjectId(), permittedAccounts, domainIdRecursiveListProject, cmd.listAll(), false);
+        Long domainId = domainIdRecursiveListProject.first();
+        Boolean isRecursive = domainIdRecursiveListProject.second();
+        ListProjectResourcesCriteria listProjectResourcesCriteria = domainIdRecursiveListProject.third();
```
Comment (anshul1886) [inline server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: @nvazquez What is changed here?

File: api/src/org/apache/cloudstack/api/command/user/snapshot/ListSnapshotsCmd.java [hunk 4]
```diff
@@ -115,8 +115,10 @@ public void execute() {
         List<SnapshotResponse> snapshotResponses = new ArrayList<SnapshotResponse>();
         for (Snapshot snapshot : result.first()) {
             SnapshotResponse snapshotResponse = _responseGenerator.createSnapshotResponse(snapshot);
-            snapshotResponse.setObjectName("snapshot");
-            snapshotResponses.add(snapshotResponse);
+            if (snapshotResponse != null) {
+                snapshotResponse.setObjectName("snapshot");
```
Comment (anshul1886) [inline api/src/org/apache/cloudstack/api/command/user/snapshot/ListSnapshotsCmd.java]: Is any issue observed here for which null check is added?

File: server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java (hunk 3 above)
Comment (nvazquez) [inline server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: Hi @anshul1886, at the beggining of each line there's one space added, to be consistent with indentation

File: api/src/org/apache/cloudstack/api/command/user/snapshot/ListSnapshotsCmd.java (hunk 4 above)
Comment (nvazquez) [inline api/src/org/apache/cloudstack/api/command/user/snapshot/ListSnapshotsCmd.java]: This check is added as we introduced a case in `createSnapshotResponse` null is returned, and in that case we should skip the response

Comment (anshul1886) [inline api/src/org/apache/cloudstack/api/command/user/snapshot/ListSnapshotsCmd.java]: What is the use case for null case when that can happen?

File: server/src/com/cloud/api/ApiResponseHelper.java [hunk 5]
```diff
@@ -526,16 +529,18 @@ public static DataStoreRole getDataStoreRole(Snapshot snapshot, SnapshotDataStor
         }
 
         long storagePoolId = snapshotStore.getDataStoreId();
-        DataStore dataStore = dataStoreMgr.getDataStore(storagePoolId, DataStoreRole.Primary);
+        if (snapshotStore.getState() != null && ! snapshotStore.getState().equals(ObjectInDataStoreStateMachine.State.Destroyed)) {
```
Comment (anshul1886) [inline server/src/com/cloud/api/ApiResponseHelper.java]: As pointed out by @mike-tutkowski store cannot be destroyed if there are snapshots. Also that enum is meant for snapshot (and entities like that) and not for snapshot store.

File: server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java (hunk 3 above)
Comment (anshul1886) [inline server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java]: These type of changes should not be done on unrelated file as it makes following changes difficult.

File: server/src/com/cloud/api/ApiResponseHelper.java (hunk 5 above)
Comment (mike-tutkowski) [inline server/src/com/cloud/api/ApiResponseHelper.java]: For managed storage, that is true. For unmanaged storage, one should be able to delete the primary storage from which a snapshot was created and the snapshot should be able to remain (since the snapshot, in this case, is on secondary storage).

**Comment by serg38** [comment]
@mike-tutkowski @anshul1886 @rhtyd We had enough LGTM and successful test results for this PR. Are there any concern before merging it?

**Comment by rhtyd** [comment]
@serg38 sure
@blueorangutan package

**Comment by mike-tutkowski** [comment]
@serg38 @anshul1886 @rhtyd Maybe someone could take a look at my comments before we check this in and see if they think any of those should be implemented here. Thanks

**Comment by rhtyd** [comment]
@mike-tutkowski sorry I don't have the bandwidth for a review, though I can facilitate with running tests. Maybe @anshul1886 and/or @serg38 can comment. Thanks.

@blueorangutan test

**Comment by serg38** [comment]
@mike-tutkowski The only comments in this PR are about deletion of PS with a snapshot copy. I am 100% agree with your statement. This PR excludes destroyed snapshot copies including ones on the PS from evaluation during the listSnaphsot call thus avoiding NPE if original PS is already deleted.
Did you have other suggestions for this PR?

**Comment by anshul1886** [comment]
@serg38 I had some open questions on code which are not yet answered specifically related to use case of null case and comparison of non comparable properties. 

Also, For unmanaged storage there is nothing called snapshot copy on primary yet once the snapshot is created. So I am not understanding what are you referring to here. I had replied to your comment on dev mail but didn't get any response.

**Comment by nvazquez** [comment]
@mike-tutkowski @anshul1886 thanks for your comments! I'll work on them!

**Comment by nvazquez** [comment]
Closing this PR as #1847 includes this changes in a simpler way
